packages/core/saved-objects/core-saved-objects-server/src/model_version.ts

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { SavedObjectMigrationContext } from './migration';
+import type { SavedObjectUnsanitizedDoc } from './serialization';
+import type { SavedObjectsMappingProperties } from './mapping_definition';
+
+/**
+ * Root structure defining a model version for a given saved object type.
+ *
+ * model version are mostly described by the operations necessary to evolve the model from the prior to the next
+ * version, and going through the temporary version.
+ */
+export interface SavedObjectsModelVersion<
+  PreviousAttrs = unknown,
+  CompatAttrs = unknown,
+  NextAttrs = unknown
+> {
+  expand: SavedObjectsModelVersionExpand<PreviousAttrs, CompatAttrs>;
+  contract: SavedObjectsModelVersionContract<CompatAttrs, NextAttrs>;
+}
+
+export interface SavedObjectsModelVersionExpand<PreviousAttrs = unknown, CompatAttrs = unknown> {
+  /**
+   * An (optional) list of new fields that will be added to the mapping during the expand phase.
+   */
+  addedMappings?: SavedObjectsMappingProperties;
+  /**
+   * An (optional) bidirectional migration to migrate the data from the previous to the compatibility state.
+   */
+  migration?: SavedObjectModelMigration<PreviousAttrs, CompatAttrs>;
+}
+
+export interface SavedObjectsModelVersionContract<CompatAttrs = unknown, NextAttrs = unknown> {
+  /**
+   * An (optional) list of paths to mappings that needs to be deleted during the contract phase.
+   */
+  removedMappings?: string[];
```

Comment:
The list of fields that needs to be removed during the contract phase. I chose to go with paths, because defining removals with a full mapping definition is awkward (and we only need the names/paths of the fields to remove)